Document the Rule Query Inspector for threshold rules (#6746)
elastic/docs-content: explore-analyze/alerting/alerts/inspect-rule-queries.md

---
navigation_title: Inspect rule queries

> description: Use the rule query inspector to view the Elasticsearch request behind a rule, confirm it targets the right data, and diagnose why an alert was or wasn't generated.

Reviewer: ```suggestion
description: Use the rule query inspector to view the Elasticsearch request a rule used, confirm it targeted the right data, and diagnose why alerts were or weren't generated.
```

applies_to:
  stack: ga 9.5
  serverless: ga
products:
  - id: kibana
---

# Inspect rule queries [inspect-rule-queries]

> The rule query inspector lets you view the {{es}} request that a rule sends when it evaluates your data. Use it to understand the query structure, confirm the rule is targeting the right data, and diagnose why an alert was or wasn't generated.

Reviewer: ```suggestion
The rule query inspector lets you view the {{es}} request that a rule sent when it evaluated your data. Use it to understand the query structure, confirm the rule targeted the right data, and diagnose why an alert was or wasn't generated.
```
I might be missing something but as I said in an earlier comment that's lower on the diff, it seems like this is historical, right, so past tense?

::::{note}
:applies_to: {"stack": "ga 9.5", "serverless": "ga"}
Currently, the rule query inspector is only available for **custom threshold rules**. 
::::

## Access the inspector [inspect-access]

The inspector is available from two places, each showing a different query:

> * **Rule details page**: Open **{{stack-manage-app}}** > **{{rules-ui}}**, find your rule, and click its name to open the rule details page. Click **Rule query inspector**. The inspector builds the query from the rule's _current_ parameters. Use this view to verify that the rule is configured correctly and would match the data you expect.

Reviewer: ```suggestion
* **Rule details page**: Open **{{stack-manage-app}}** > **{{rules-ui}}**, find your rule, and click its name to open its details page. Click **Rule query inspector**. The inspector builds the query from the rule's _current_ parameters. Use this view to verify that the rule is configured correctly and would match the data you expect.
```

* **Alert details page**: Go to the **Alerts** page, then open an individual alert. Click **Rule query inspector**. The inspector uses the rule parameters _as they existed when that specific alert was generated_, including the exact evaluation time range. Use this view to understand why a particular alert was or wasn't triggered.

> The key difference is that the rule details page reflects the rule as it is _now_, while the alert details page reflects the rule as it was _then_. If you've edited the rule since an alert was generated, the two inspectors will show different queries.

Reviewer: Okay, I see that it actually does both! Super cool. This seems important. I wonder if this is worth mentioning in the intro section rather than waiting for the "Access your inspector" section

Author: Good idea! Bumping it up could help with framing the feature so users have a better understanding before diving into the rest of the page.

Author: Moved in [`4995d81` (this PR)](https://github.com/elastic/docs-content/pull/6746/commits/4995d816b7a8da392fce4b014458f1548448e2a2).

## What the inspector shows [inspect-tabs]

> The inspector displays the {{es}} query the rule sent, the raw response it received, and how long the query took to execute.

Reviewer: Tricky. Because sometimes is displays the query the rule _will_ send, vs. _sent_. What about just: 
```suggestion
The inspector displays the {{es}} query associated with the rule, the most recent raw response it received, and how long the query took to run.
```

Author: Looks good, just making one minor tweak:

```suggestion
The inspector displays the {{es}} query made by the rule, the most recent raw response the rule received, and how long the query took to run.
```

| Element | Description |
| --- | --- |
| **Criterion dropdown** | Appears when a rule has multiple criteria. Each entry is labeled with its criterion number and metric (for example, `Criterion 1: avg(system.cpu.total.norm.pct)`). Selecting a criterion updates both the **Request** and **Response** tabs to show the query and results for that specific condition. |
| **Request** | Shows the full {{es}} query that the rule sends when it evaluates your data. Use it to verify the index pattern, time range, query filter, and aggregations match what you configured in the rule. |
| **Response** | Shows the raw {{es}} response. Use it to confirm whether data was found, whether the groups you expect are present, and what values the rule was working with when it made its alerting decision. |
| **Request time** | Shows how long {{es}} took to execute the query. This measures the query portion of rule execution only. It doesn't include time spent waiting in the task queue or processing actions after the query returns. Use it to identify whether the query itself is the bottleneck when a rule is slow. |

> ## Factors that affect request time [inspect-request-time-factors]

Reviewer: I'm unclear here on whether this is specifically about the efficiency of getting Rule Query Inspector data, or the efficiency of detection rule queries. I think the latter, but then I'm not quite sure why it's on this page because it seems like more general detection rule content than the rule query inspector. Maybe there's room a tutorial doc about how to use the Inspector to assess the efficiency of your detection rule queries? Or maybe an intro to this section would help me understand why it belongs here

Author: That's a fair point. This table is almost repeating the info in the "Using the inspector" section, just in a different way. I'll remove it for now and add it elsewhere if I find a good spot. Thanks for flagging this. : )

Author: Removed in [`4995d81` (this PR)](https://github.com/elastic/docs-content/pull/6746/commits/4995d816b7a8da392fce4b014458f1548448e2a2).
